feat: implements ARRAY_JOIN as requested in (#5028)

hpgrahsl opened the pull request:
### Description 
This PR implements the UDF / scalar function ARRAY_JOIN as requested in (#5028). It creates a flat string representation of all the elements contained in the given array.

### Testing done 
There are unit tests with several applications of the UDF to verify that it follows the expected behaviour as discussed in the linked issue.

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayJoin.java:
```diff
@@ -0,0 +1,111 @@
+public class ArrayJoin {
+
+  private static final String DEFAULT_DELIMITER = ",";
+  private static final Set<Class> KSQL_PRIMITIVES = new HashSet<>(
+      Arrays.asList(Boolean.class,Integer.class,BigInteger.class,Double.class,String.class)
+  );
```
nit: let's use `ImmutableSet.of`

hpgrahsl replied:
ok. will do.

hpgrahsl replied:
done!

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayJoin.java:
```diff
@@ -0,0 +1,111 @@
+
+  private static final String DEFAULT_DELIMITER = ",";
+  private static final Set<Class> KSQL_PRIMITIVES = new HashSet<>(
+      Arrays.asList(Boolean.class,Integer.class,BigInteger.class,Double.class,String.class)
```
why `BigInteger`? I think ksqlDB only uses `BigDecimal` internally - there is also `Long.class` if you are referring to the sql `BIGINT` type

hpgrahsl replied:
fair point. I guess I mistakenly assumed that sql `BIGINT` would map to `BigInteger` which obviously is not the case? Should I switch to `Long` cause `BigDecimal` seems even weirder to me than `BigInteger` when thinking about it.

agavra replied:
I was a little unclear; we should remove `BigInteger` and add both `Long` and `BigDecimal`. `BigDecimal` is the java class that we use to represent the sql `DECIMAL` type, which should be supported in this method

hpgrahsl replied:
Thx for clarifying. The reason I didn't consider DECIMAL is because it simply didn't work for me. Maybe I'm too stupid but trying to use the ksql CLI and create an `ARRAY` of `DECIMAL`s   never worked for me thus I thought it's currently not supported :)

agavra replied:
Not stupid, haha - our UX isn't great with decimals, but the below should work:
```sql
ksql> CREATE STREAM dec_array (val ARRAY<DECIMAL(2,1)>) WITH (kafka_topic='foo', value_format='json', partitions=1);

 Message
----------------
 Stream created
----------------
ksql> INSERT INTO dec_array (val) VALUES (ARRAY[2.1]);
ksql> SELECT * FROM dec_array EMIT CHANGES;
+----------------------------------------------+----------------------------------------------+
|ROWKEY                                        |VAL                                           |
+----------------------------------------------+----------------------------------------------+
|null                                          |[2.1]                                         |
```

hpgrahsl replied:
done!

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayJoin.java:
```diff
@@ -0,0 +1,111 @@
+      final String delimiter
+  ) {
+
+    if (array == null) {
```
what should the behavior be on a `null` delimiter? should we default to empty string delimiter or just return `null`? cc @derekjn

hpgrahsl replied:
I have no preference but you are right that it has to be specified by someone ;-)

derekjn replied:
I think that interpreting a `null` delimiter as the empty string makes sense here.

hpgrahsl replied:
done!

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayJoin.java:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  private static void handleListType(final List<?> element, final StringJoiner joiner) {
+    element.forEach(e -> processElement(e, joiner));
```
if we have a list of list, should it flatten the nested types or just return a string representation of the element? it seems odd to me that `ARRAY[ARRAY[1, 2], ARRAY[3]]` would flatten to the same result as `ARRAY[1, 2, 3]`

same concern for map and struct - I think my preference would be to support only primitives in the first implementation. Other languages have similar restrictions, e.g. python (which is even more restrictive):
```python
>>> '.'.join([['1', '2']])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sequence item 0: expected string, list found
```

hpgrahsl replied:
There were no clear indications about this in the issue discussion which is why I implemented it in this way, which is IMHO a "natural" choice. Remember that the original idea was that the result is just a flat string. If you want to see a different string result for the example you are referring to it essentially means to introduce further delimiters which can capture the semantics of nesting which might be out of scope for such a function. If someone really wants to have that I think it's better to decide for other forms of encoding e.g. a fully-fledged JSON-string representation, but I guess that's a different story.

agavra replied:
cc @derekjn @rmoff to chime in here, I'm happy with whatever we settle on

derekjn replied:
I don't think we should implicitly flatten anything here. If I passed an array of arrays, I'd expect the result to be the string representation of each array, joined together by the delimiter. If that adds too much complexity to the implementation then I agree with @agavra that it would be reasonable to require primitive types, at least for the initial implementation. The semantics around concatenating string representations of structured types (especially if they have nondeterministic string reps) can start to get pretty messy.

hpgrahsl replied:
hi @derekjn so do I get you right that you want to capture the nesting semantics somehow in the resulting string representation? if so how _exactly_ should this look like? it would be helpful if you can provide to or some examples at least one for arrays of arrays, maps of maps or nested structs.

based on that we can then decide if the implementation should be done or we simply drop support for anything but flat arrays containing primitives. what I still quite don't get is why you have your issues with the current way it's implemented. think about a simple CSV-like string - in a basic CSV encoding you would end up with delimiting elements as is currently done since you don't capture complex structure in a simple CSV-style string.

derekjn replied:
For all of those examples I think we'd just want to use whatever `CAST(value AS STRING)` would return for the array element. That being said, I'm fine with only supporting primitive types for the initial implementation as @agavra suggested. Let's just go with that.

hpgrahsl replied:
dropped support for complex nested types for this initial version of ARRAY_JOIN

agavra commented on ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayJoinTest.java:
```diff
@@ -0,0 +1,258 @@
+
+package io.confluent.ksql.function.udf.array;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
```
nit: we prefer hamcrest asserts in our tests, you can look at other tests for examples of `assertThat(value, matches(matcher))`

hpgrahsl replied:
sure I can do that.
re-nit from my side: however, when it comes to consistency of the way things are done in existing UDFs we could discuss days if not weeks ;-)

hpgrahsl replied:
switched to hamcrest matchers

agavra commented on ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayJoinTest.java:
```diff
@@ -0,0 +1,258 @@
+  }
+
+  @Test
+  public void shouldReturnCorrectStringForFlatArraysWithPrimitiveTypes() {
```
can you add decimal and long types into this? (see not above that `BigInteger` and `Long` are not the same in java)

hpgrahsl replied:
sure. but both or either of them? see my question above at the `BigInteger` comment.

hpgrahsl replied:
done!

hpgrahsl commented:
> It is missing what we call "QueryTranslationTests", you can see an example of one in `substring.json`

I have to look into this and might have further questions most likely about this. I wrote the tests in relation to other, existing array function tests and haven't seen anything like this.

JimGalasyn commented on docs/developer-guide/ksqldb-reference/scalar-functions.md:
````diff
@@ -221,6 +221,18 @@ SLICE(col1, from, to)
+ARRAY_JOIN(col1, delimiter)
+```
+
+Creates a flat String representation of all the elements contained in the given array.
````
```suggestion
Creates a flat string representation of all the elements contained in the given array.
```

hpgrahsl replied:
THX. will be changed accordingly.

hpgrahsl replied:
done!

JimGalasyn commented on docs/developer-guide/ksqldb-reference/scalar-functions.md:
````diff
@@ -221,6 +221,18 @@ SLICE(col1, from, to)
+```
+
+Creates a flat String representation of all the elements contained in the given array.
+The elements in the resulting String are separated by the chosen `delimiter`, 
````
```suggestion
The elements in the resulting string are separated by the chosen `delimiter`, 
```

hpgrahsl replied:
THX. will be changed accordingly.

hpgrahsl replied:
done!

JimGalasyn commented on docs/developer-guide/ksqldb-reference/scalar-functions.md:
```diff
@@ -221,6 +221,18 @@ SLICE(col1, from, to)
+Creates a flat String representation of all the elements contained in the given array.
+The elements in the resulting String are separated by the chosen `delimiter`, 
+which is an optional parameter that falls back to a comma `,`. The array may contain
+all valid ksql types and supports nested and complex data structures as well, 
```
```suggestion
all valid ksqlDB types and supports nested and complex data structures as well, 
```

hpgrahsl replied:
THX. will be changed accordingly.

hpgrahsl replied:
done!

hpgrahsl commented:
as requested by reviewers:

- removed BigInteger from primitive types
- added Long and BigDecimal to supported primitive types
- dropped support for complex nested types
- switched to hamcrest matchers in unit tests
- added QueryTranslationTest for via PlannedTestGeneratorTest
- adapted docs accordingly
